find_broken_links.py

```diff
@@ -0,0 +1,118 @@
+import re
+import socket
+import sys
+import urllib.request
+from glob import iglob
+from typing import Iterator, TextIO, Tuple
+from urllib.error import HTTPError, URLError
+
+from dataclasses import dataclass
+
+REQUEST_HEADERS = {'User-Agent': 'https://github.com/RTBHOUSE/here-be-pythons'}
+TIMEOUT_SECONDS = 5
+URL_REGEX = re.compile(
+    r'(?:http|ftp|https)://'                    # protocol
+    r'(?:[\w_-]+(?:(?:\.[\w_-]+)+))'            # domain
+    r'(?:[\w.,@?^=%&:/~+#-]*[\w@?^=%&/~+#-])?'  # path
+)  # yapf: disable
+
+
+def find_broken_links(files_glob_pattern='docs/source/**/*.rst') -> None:
+    counters = Counters()
+
+    for file in _yield_files(files_glob_pattern):
+        _process_urls_from_file(file, counters)
+
+    _print_report(counters)
+
+    if counters.broken_links_num == 0:
+        sys.exit(0)
+    else:
+        sys.exit(1)
+
+
+@dataclass
+class Link:
+    url: str
+    health: str = None
+    status_code: int = None
+
+
+@dataclass
+class Counters:
+    ok_links_num: int = 0
+    broken_links_num: int = 0
+    timeout_links_num: int = 0
+
+
+def _yield_files(files_glob_pattern: str) -> Iterator[TextIO]:
+    files_paths = iglob(files_glob_pattern)
+    for file_path in files_paths:
+        with open(file_path) as file:
+            yield file
+
+
+def _process_urls_from_file(file: TextIO, counters: Counters) -> None:
+    for link, line_num in _yield_links_from_file(file):
+        _update_link_health(link, counters)
+        if link.health == 'broken':
+            print(f'url:       {link.url}')
+            print(f'returns:   {link.status_code}')
+            print(f'file:      {file.name}:{line_num}')
+            print('')
+
+
+def _yield_links_from_file(file: TextIO) -> Iterator[Tuple[Link, int]]:
+    for line_num, line in enumerate(file, start=1):
+        urls = URL_REGEX.findall(line)
+        for url in urls:
+            yield Link(url), line_num
+
+
+def _update_link_health(link: Link, counters: Counters) -> None:
+    request = urllib.request.Request(link.url, headers=REQUEST_HEADERS)
+
+    try:
+        urllib.request.urlopen(request, timeout=TIMEOUT_SECONDS)
+        link.health = 'ok'
+        counters.ok_links_num += 1
+
+    except (HTTPError, URLError) as exc:
```

Comment:
a co z ContentTooShortError?

Reply:
My intention is to keep the code simple and handle only the exceptions that occurred so far. I even **want** the script to crash in such cases in order to study them dilligently.

I apply this logic, because this is a helper/utility feature for checking broken links - nothing more, nothing less :)

Reply:
Understood. 
Don't you think, though, that if you handle two descriptive errors for lib used you probably should at least give a thought to the remaining one(s)?

Reply:
Well, agree to giving it a thought, yet without handling (at least in this case). I follow YAGNI whenever possible. 

As mentioned, I want the script to crash to better understand what happened - instead of writing up some logic before it is needed. Basing on what I have seen so far, there are lots and lots of places in our code in which we do not cover every single case that we assume may occurr, and this is for a reason of avoiding unnecesary work. 

To stress it once again, this is a helper function, not a web server intended to survive almost every possible event!